File: tests/python/unit/test_unit_runner.py
#!/usr/bin/env python3
"""
Unit test runner for QtForge Python bindings.
Runs all unit tests with comprehensive reporting and coverage analysis.
"""

import pytest
import sys
import os
import time
import subprocess
from pathlib import Path
from typing import List, Dict, Any

# Add the build directory to Python path for testing
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../../build'))

try:
    import qtforge
    BINDINGS_AVAILABLE = True
    print("✅ QtForge bindings are available")
except ImportError as e:
    BINDINGS_AVAILABLE = False
    print(f"❌ QtForge bindings not available: {e}")


class UnitTestRunner:
    """Comprehensive unit test runner for QtForge Python bindings."""
    
    def __init__(self):
        self.test_dir = Path(__file__).parent
        self.test_modules = [
            "test_core_unit.py",
            "test_utils_unit.py", 
            "test_communication_unit.py",
            "test_security_unit.py",
            "test_managers_unit.py",
            "test_orchestration_unit.py",
            "test_monitoring_unit.py",
            "test_transactions_unit.py",
            "test_composition_unit.py",
            "test_marketplace_unit.py",
            "test_threading_unit.py"
        ]
        self.results = {}
    
    def check_test_files(self) -> Dict[str, bool]:
        """Check which test files exist."""
        existing_tests = {}
        missing_tests = []
        
        for test_module in self.test_modules:
            test_path = self.test_dir / test_module
            exists = test_path.exists()
            existing_tests[test_module] = exists
            if not exists:
                missing_tests.append(test_module)
        
        if missing_tests:
            print("⚠️  Some unit test modules are missing:")
            for missing in missing_tests:
                print(f"   - {missing}")
            print()
        
        return existing_tests
    
    def run_individual_test(self, test_file: str) -> Dict[str, Any]:
        """Run an individual test file and return results."""
        print(f"🧪 Running {test_file}...")
        
        test_path = self.test_dir / test_file
        if not test_path.exists():
            return {
                'status': 'skipped',
                'reason': 'File not found',
                'duration': 0,
                'tests_run': 0,
                'failures': 0,
                'errors': 0
            }
        
        start_time = time.time()
        
        # Run pytest on the specific file
        pytest_args = [
            str(test_path),
            "-v",
            "--tb=short",
            "--strict-markers",
            "-x",  # Stop on first failure for individual tests
            "--durations=5",  # Show 5 slowest tests
            "-q"  # Quiet output for cleaner results
        ]
        
        try:
            result = pytest.main(pytest_args)
            end_time = time.time()
            duration = end_time - start_time
            
            # Parse pytest exit codes
            if result == 0:
                status = 'passed'
            elif result == 1:
                status = 'failed'
            elif result == 2:
                status = 'interrupted'
            elif result == 3:
                status = 'internal_error'
            elif result == 4:
                status = 'usage_error'
            elif result == 5:
                status = 'no_tests'
            else:
                status = 'unknown'
            
            return {
                'status': status,
                'duration': duration,
                'exit_code': result
            }
            
        except Exception as e:
            end_time = time.time()
            duration = end_time - start_time
            
            return {
                'status': 'error',
                'duration': duration,
                'error': str(e)
            }
    
    def run_all_tests(self) -> Dict[str, Any]:
        """Run all available unit tests."""
        if not BINDINGS_AVAILABLE:
            print("❌ Cannot run unit tests - QtForge bindings not available")
            return {'status': 'skipped', 'reason': 'Bindings not available'}
        
        print("🚀 Running QtForge Python unit tests...")
        print("=" * 60)
        
        existing_tests = self.check_test_files()
        available_tests = [test for test, exists in existing_tests.items() if exists]
        
        if not available_tests:
            print("❌ No unit test files found!")
            return {'status': 'no_tests'}
        
        print(f"📋 Found {len(available_tests)} unit test modules")
        print()
        
        overall_start_time = time.time()
        results = {}
        
        # Run each test module
        for test_file in available_tests:
            result = self.run_individual_test(test_file)
            results[test_file] = result
            
            # Print immediate result
            status = result['status']
            duration = result.get('duration', 0)
            
            if status == 'passed':
                print(f"   ✅ {test_file} - PASSED ({duration:.2f}s)")
            elif status == 'failed':
                print(f"   ❌ {test_file} - FAILED ({duration:.2f}s)")
            elif status == 'skipped':
                print(f"   ⏭️  {test_file} - SKIPPED ({result.get('reason', 'Unknown')})")
            else:
                print(f"   ⚠️  {test_file} - {status.upper()} ({duration:.2f}s)")
        
        overall_end_time = time.time()
        total_duration = overall_end_time - overall_start_time
        
        # Calculate summary statistics
        passed = sum(1 for r in results.values() if r['status'] == 'passed')
        failed = sum(1 for r in results.values() if r['status'] == 'failed')
        skipped = sum(1 for r in results.values() if r['status'] == 'skipped')
        errors = sum(1 for r in results.values() if r['status'] == 'error')
        
        print()
        print("=" * 60)
        print(f"📊 Unit Test Summary:")
        print(f"   Total modules: {len(available_tests)}")
        print(f"   Passed: {passed}")
        print(f"   Failed: {failed}")
        print(f"   Skipped: {skipped}")
        print(f"   Errors: {errors}")
        print(f"   Total time: {total_duration:.2f}s")
        
        if failed == 0 and errors == 0:
            print("✅ All unit tests passed!")
            overall_status = 'passed'
        else:
            print("❌ Some unit tests failed!")
            overall_status = 'failed'
        
        return {
            'status': overall_status,
            'total_duration': total_duration,
            'results': results,
            'summary': {
                'total': len(available_tests),
                'passed': passed,
                'failed': failed,
                'skipped': skipped,
                'errors': errors
            }
        }
    
    def run_coverage_analysis(self) -> Dict[str, Any]:
        """Run coverage analysis on unit tests."""
        print("📈 Running coverage analysis...")
        
        try:
            # Try to run with coverage
            coverage_args = [
                sys.executable, "-m", "coverage", "run",
                "--source=qtforge",
                "-m", "pytest",
                str(self.test_dir),
                "-v", "--tb=short"
            ]
            
            result = subprocess.run(coverage_args, capture_output=True, text=True)
            
            if result.returncode == 0:
                # Generate coverage report
                report_args = [sys.executable, "-m", "coverage", "report"]
                report_result = subprocess.run(report_args, capture_output=True, text=True)
                
                if report_result.returncode == 0:
                    print("Coverage Report:")
                    print(report_result.stdout)
                    return {'status': 'success', 'report': report_result.stdout}
                else:
                    print(f"Coverage report failed: {report_result.stderr}")
                    return {'status': 'report_failed', 'error': report_result.stderr}
            else:
                print(f"Coverage run failed: {result.stderr}")
                return {'status': 'run_failed', 'error': result.stderr}
                
        except FileNotFoundError:
            print("⚠️  Coverage tool not available. Install with: pip install coverage")
            return {'status': 'tool_not_available'}
        except Exception as e:
            print(f"Coverage analysis error: {e}")
            return {'status': 'error', 'error': str(e)}
    
    def generate_test_report(self, results: Dict[str, Any]) -> str:
        """Generate a detailed test report."""
        report_lines = [
            "QtForge Python Bindings Unit Test Report",
            "=" * 50,
            f"Generated: {time.strftime('%Y-%m-%d %H:%M:%S')}",
            ""
        ]
        
        if results['status'] == 'skipped':
            report_lines.extend([
                "❌ Tests were skipped:",
                f"   Reason: {results.get('reason', 'Unknown')}",
                ""
            ])
            return "\n".join(report_lines)
        
        summary = results.get('summary', {})
        report_lines.extend([
            "📊 Summary:",
            f"   Total test modules: {summary.get('total', 0)}",
            f"   Passed: {summary.get('passed', 0)}",
            f"   Failed: {summary.get('failed', 0)}",
            f"   Skipped: {summary.get('skipped', 0)}",
            f"   Errors: {summary.get('errors', 0)}",
            f"   Total duration: {results.get('total_duration', 0):.2f}s",
            ""
        ])
        
        # Detailed results
        test_results = results.get('results', {})
        if test_results:
            report_lines.extend([
                "📋 Detailed Results:",
                ""
            ])
            
            for test_file, result in test_results.items():
                status = result['status']
                duration = result.get('duration', 0)
                
                status_icon = {
                    'passed': '✅',
                    'failed': '❌',
                    'skipped': '⏭️',
                    'error': '⚠️'
                }.get(status, '❓')
                
                report_lines.append(f"   {status_icon} {test_file}: {status.upper()} ({duration:.2f}s)")
                
                if 'error' in result:
                    report_lines.append(f"      Error: {result['error']}")
                if 'reason' in result:
                    report_lines.append(f"      Reason: {result['reason']}")
        
        return "\n".join(report_lines)


def main():
    """Main entry point for unit test runner."""
    runner = UnitTestRunner()
    
    # Parse command line arguments
    if len(sys.argv) > 1:
        arg = sys.argv[1]
        
        if arg in ["-h", "--help"]:
            print("QtForge Python Unit Test Runner")
            print()
            print("Usage:")
            print("  python test_unit_runner.py                # Run all unit tests")
            print("  python test_unit_runner.py --coverage     # Run with coverage analysis")
            print("  python test_unit_runner.py --report       # Generate detailed report")
            print("  python test_unit_runner.py --help         # Show this help")
            return 0
        
        elif arg == "--coverage":
            # Run tests with coverage
            results = runner.run_all_tests()
            if results['status'] != 'skipped':
                runner.run_coverage_analysis()
            return 0 if results['status'] == 'passed' else 1
        
        elif arg == "--report":
            # Run tests and generate detailed report
            results = runner.run_all_tests()
            report = runner.generate_test_report(results)
            
            # Save report to file
            report_file = runner.test_dir / "unit_test_report.txt"
            with open(report_file, 'w') as f:
                f.write(report)
            
            print(f"\n📄 Detailed report saved to: {report_file}")
            return 0 if results['status'] == 'passed' else 1
    
    # Default: run all tests
    results = runner.run_all_tests()
    return 0 if results['status'] == 'passed' else 1


if __name__ == "__main__":
    sys.exit(main())

File: tests/lua/unit/test_unit_runner.lua
#!/usr/bin/env lua
--[[
Unit test runner for QtForge Lua bindings.
Runs all unit tests with comprehensive reporting and coverage analysis.
]]

-- Test runner class
local TestRunner = {}
TestRunner.__index = TestRunner

function TestRunner:new()
    local obj = {
        test_dir = ".",
        test_modules = {
            "test_core_unit.lua",
            "test_utils_unit.lua",
            "test_communication_unit.lua",
            "test_security_unit.lua",
            "test_managers_unit.lua",
            "test_orchestration_unit.lua",
            "test_monitoring_unit.lua",
            "test_transactions_unit.lua",
            "test_composition_unit.lua",
            "test_marketplace_unit.lua",
            "test_threading_unit.lua"
        },
        results = {}
    }
    setmetatable(obj, TestRunner)
    return obj
end

function TestRunner:check_qtforge_availability()
    local success, qtforge = pcall(require, "qtforge")
    if success then
        print("✅ QtForge Lua bindings are available")
        return true, qtforge
    else
        print(string.format("❌ QtForge Lua bindings not available: %s", qtforge))
        return false, nil
    end
end

function TestRunner:check_test_files()
    local existing_tests = {}
    local missing_tests = {}
    
    for _, test_module in ipairs(self.test_modules) do
        local file = io.open(test_module, "r")
        if file then
            file:close()
            table.insert(existing_tests, test_module)
        else
            table.insert(missing_tests, test_module)
        end
    end
    
    if #missing_tests > 0 then
        print("⚠️  Some unit test modules are missing:")
        for _, missing in ipairs(missing_tests) do
            print(string.format("   - %s", missing))
        end
        print()
    end
    
    return existing_tests, missing_tests
end

function TestRunner:run_individual_test(test_file)
    print(string.format("🧪 Running %s...", test_file))
    
    local start_time = os.clock()
    
    -- Execute the test file
    local success, result = pcall(function()
        return dofile(test_file)
    end)
    
    local end_time = os.clock()
    local duration = end_time - start_time
    
    if success then
        local exit_code = result or 0
        if exit_code == 0 then
            return {
                status = 'passed',
                duration = duration,
                exit_code = exit_code
            }
        else
            return {
                status = 'failed',
                duration = duration,
                exit_code = exit_code
            }
        end
    else
        return {
            status = 'error',
            duration = duration,
            error = tostring(result)
        }
    end
end

function TestRunner:run_all_tests()
    local available, qtforge = self:check_qtforge_availability()
    if not available then
        print("❌ Cannot run unit tests - QtForge bindings not available")
        return {status = 'skipped', reason = 'Bindings not available'}
    end
    
    print("🚀 Running QtForge Lua unit tests...")
    print(string.rep("=", 60))
    
    local existing_tests, missing_tests = self:check_test_files()
    
    if #existing_tests == 0 then
        print("❌ No unit test files found!")
        return {status = 'no_tests'}
    end
    
    print(string.format("📋 Found %d unit test modules", #existing_tests))
    print()
    
    local overall_start_time = os.clock()
    local results = {}
    
    -- Run each test module
    for _, test_file in ipairs(existing_tests) do
        local result = self:run_individual_test(test_file)
        results[test_file] = result
        
        -- Print immediate result
        local status = result.status
        local duration = result.duration
        
        if status == 'passed' then
            print(string.format("   ✅ %s - PASSED (%.2fs)", test_file, duration))
        elseif status == 'failed' then
            print(string.format("   ❌ %s - FAILED (%.2fs)", test_file, duration))
        elseif status == 'error' then
            print(string.format("   ⚠️  %s - ERROR (%.2fs)", test_file, duration))
            if result.error then
                print(string.format("      Error: %s", result.error))
            end
        else
            print(string.format("   ❓ %s - %s (%.2fs)", test_file, status:upper(), duration))
        end
    end
    
    local overall_end_time = os.clock()
    local total_duration = overall_end_time - overall_start_time
    
    -- Calculate summary statistics
    local passed = 0
    local failed = 0
    local errors = 0
    
    for _, result in pairs(results) do
        if result.status == 'passed' then
            passed = passed + 1
        elseif result.status == 'failed' then
            failed = failed + 1
        elseif result.status == 'error' then
            errors = errors + 1
        end
    end
    
    print()
    print(string.rep("=", 60))
    print("📊 Unit Test Summary:")
    print(string.format("   Total modules: %d", #existing_tests))
    print(string.format("   Passed: %d", passed))
    print(string.format("   Failed: %d", failed))
    print(string.format("   Errors: %d", errors))
    print(string.format("   Total time: %.2fs", total_duration))
    
    local overall_status
    if failed == 0 and errors == 0 then
        print("✅ All unit tests passed!")
        overall_status = 'passed'
    else
        print("❌ Some unit tests failed!")
        overall_status = 'failed'
    end
    
    return {
        status = overall_status,
        total_duration = total_duration,
        results = results,
        summary = {
            total = #existing_tests,
            passed = passed,
            failed = failed,
            errors = errors
        }
    }
end

function TestRunner:generate_test_report(results)
    local report_lines = {
        "QtForge Lua Bindings Unit Test Report",
        string.rep("=", 50),
        string.format("Generated: %s", os.date("%Y-%m-%d %H:%M:%S")),
        ""
    }
    
    if results.status == 'skipped' then
        table.insert(report_lines, "❌ Tests were skipped:")
        table.insert(report_lines, string.format("   Reason: %s", results.reason or "Unknown"))
        table.insert(report_lines, "")
        return table.concat(report_lines, "\n")
    end
    
    local summary = results.summary or {}
    table.insert(report_lines, "📊 Summary:")
    table.insert(report_lines, string.format("   Total test modules: %d", summary.total or 0))
    table.insert(report_lines, string.format("   Passed: %d", summary.passed or 0))
    table.insert(report_lines, string.format("   Failed: %d", summary.failed or 0))
    table.insert(report_lines, string.format("   Errors: %d", summary.errors or 0))
    table.insert(report_lines, string.format("   Total duration: %.2fs", results.total_duration or 0))
    table.insert(report_lines, "")
    
    -- Detailed results
    local test_results = results.results or {}
    if next(test_results) then
        table.insert(report_lines, "📋 Detailed Results:")
        table.insert(report_lines, "")
        
        for test_file, result in pairs(test_results) do
            local status = result.status
            local duration = result.duration or 0
            
            local status_icon = {
                passed = '✅',
                failed = '❌',
                error = '⚠️'
            }
            
            local icon = status_icon[status] or '❓'
            table.insert(report_lines, string.format("   %s %s: %s (%.2fs)", 
                icon, test_file, status:upper(), duration))
            
            if result.error then
                table.insert(report_lines, string.format("      Error: %s", result.error))
            end
        end
    end
    
    return table.concat(report_lines, "\n")
end

function TestRunner:run_specific_module(module_name)
    local test_file = string.format("test_%s_unit.lua", module_name)
    
    local file = io.open(test_file, "r")
    if not file then
        print(string.format("❌ Test module not found: %s", test_file))
        return 1
    end
    file:close()
    
    print(string.format("🧪 Running tests for %s module...", module_name))
    
    local result = self:run_individual_test(test_file)
    
    if result.status == 'passed' then
        print("✅ Module tests passed!")
        return 0
    else
        print("❌ Module tests failed!")
        if result.error then
            print(string.format("Error: %s", result.error))
        end
        return 1
    end
end

function TestRunner:list_available_modules()
    print("Available test modules:")
    print()
    
    local existing_tests, missing_tests = self:check_test_files()
    
    if #existing_tests > 0 then
        print("✅ Available:")
        for _, test_file in ipairs(existing_tests) do
            local module_name = test_file:match("test_(.+)_unit%.lua")
            if module_name then
                print(string.format("   - %s", module_name))
            end
        end
    end
    
    if #missing_tests > 0 then
        print()
        print("❌ Missing:")
        for _, test_file in ipairs(missing_tests) do
            local module_name = test_file:match("test_(.+)_unit%.lua")
            if module_name then
                print(string.format("   - %s", module_name))
            end
        end
    end
end

-- Main function
local function main(args)
    local runner = TestRunner:new()
    
    if #args == 0 then
        -- Run all tests
        local results = runner:run_all_tests()
        return results.status == 'passed' and 0 or 1
        
    elseif args[1] == "--help" or args[1] == "-h" then
        print("QtForge Lua Unit Test Runner")
        print()
        print("Usage:")
        print("  lua test_unit_runner.lua                    # Run all tests")
        print("  lua test_unit_runner.lua <module>           # Run specific module tests")
        print("  lua test_unit_runner.lua --list             # List available modules")
        print("  lua test_unit_runner.lua --report           # Generate detailed report")
        print("  lua test_unit_runner.lua --help             # Show this help")
        print()
        print("Examples:")
        print("  lua test_unit_runner.lua core               # Test core module")
        print("  lua test_unit_runner.lua utils              # Test utils module")
        return 0
        
    elseif args[1] == "--list" then
        runner:list_available_modules()
        return 0
        
    elseif args[1] == "--report" then
        -- Run tests and generate detailed report
        local results = runner:run_all_tests()
        local report = runner:generate_test_report(results)
        
        -- Save report to file
        local report_file = "lua_unit_test_report.txt"
        local file = io.open(report_file, "w")
        if file then
            file:write(report)
            file:close()
            print(string.format("\n📄 Detailed report saved to: %s", report_file))
        else
            print("\n❌ Failed to save report file")
        end
        
        return results.status == 'passed' and 0 or 1
        
    else
        -- Run specific module
        return runner:run_specific_module(args[1])
    end
end

-- Parse command line arguments
local args = {}
for i = 1, #arg do
    table.insert(args, arg[i])
end

-- Run main function and exit with appropriate code
local exit_code = main(args)
os.exit(exit_code)
